app/com/linkedin/drelephant/exceptions/TonYExceptionFingerprinting.java

```diff
@@ -0,0 +1,278 @@
+    for (String exactExceptionRegex : ConfigurationBuilder.REGEX_FOR_EXACT_EXCEPTION_PATTERN_IN_TONY_LOGS.getValue()) {
+      exactExceptionRegexList.add(Pattern.compile(exactExceptionRegex));
+    }
+    for (Pattern exactExceptionPattern : exactExceptionRegexList) {
+      Matcher exact_exception_pattern_matcher = exactExceptionPattern.matcher(logData);
+      while (exact_exception_pattern_matcher.find()) {
+        int exceptionId = exact_exception_pattern_matcher.group(0).hashCode();
+        if (isExceptionLogDuplicate(exceptionId)) {
+          ExceptionInfo exceptionInfo = new ExceptionInfo();
+          exceptionInfo.setExceptionID(exceptionId);
+          exceptionInfo.setExceptionName(exact_exception_pattern_matcher.group(1));
+          exceptionInfo.setExceptionSource(ExceptionInfo.ExceptionSource.DRIVER);
+          exceptionInfo.setExceptionStackTrace(exact_exception_pattern_matcher.group(0));
+          logger.info("Offset exact " + exact_exception_pattern_matcher.start());
+          exceptionInfo.setExceptionTrackingURL(logLocationURL + LOG_START_OFFSET_PARAM +
+              exact_exception_pattern_matcher.start());
+          exactlyMatchingExceptionList.add(exceptionInfo);
+        }
+      }
+    }
+    return exactlyMatchingExceptionList;
+  }
+
+  private List<ExceptionInfo> filterOutPartialExceptionPattern(String logData,
+      String logLocationURL) {
+    List<ExceptionInfo> exceptionInfoList = new ArrayList<>();
+    List<Pattern> partialExceptionRegexList = new ArrayList<>();
+    for (String partialExceptionRegex : REGEX_FOR_PARTIAL_EXCEPTION_PATTERN_IN_TONY_LOGS.getValue()) {
+      partialExceptionRegex = partialExceptionRegex + String.format("(.*\n?){%d,%d}", 0, NUMBER_OF_STACKTRACE_LINE.getValue());
+      partialExceptionRegexList.add(Pattern.compile(partialExceptionRegex));
+    }
+    for (Pattern exactExceptionPattern : partialExceptionRegexList) {
+      Matcher partial_pattern_matcher = exactExceptionPattern.matcher(logData);
+      while (partial_pattern_matcher.find()) {
+        ExceptionInfo exceptionInfo = new ExceptionInfo();
+        exceptionInfo.setExceptionID(partial_pattern_matcher.group(0).hashCode());
+        exceptionInfo.setExceptionName(partial_pattern_matcher.group(0).split("\n", 2)[0]);
+        exceptionInfo.setExceptionSource(ExceptionInfo.ExceptionSource.DRIVER);
+        exceptionInfo.setExceptionStackTrace(partial_pattern_matcher.group(0));
+        logger.info("Offset partial " + partial_pattern_matcher.start());
+        logger.info("Match for partial exception pattern " + exceptionInfo.getExceptionStackTrace());
+
+        exceptionInfo.setExceptionTrackingURL(logLocationURL + LOG_START_OFFSET_PARAM +
+            partial_pattern_matcher.start());
+        exceptionInfoList.add(exceptionInfo);
+      }
+    }
+    return exceptionInfoList;
+  }
+
+  public void saveExceptionFingerprintingData() {
+    if (_exceptionInfoList != null) {
+      final String NOT_APPLICABLE = "NA";
```

Comment:
Nit : better to move Constant declaration up

Reply:
Used it only once so, kept it as a local variable.